configuration-service/common/projects_materialized_view.go

```diff
@@ -0,0 +1,208 @@
+	existingProject, err := mv.GetProject(prj.ProjectName)
+	if existingProject != nil {
+		return nil
+	}
+	err = mv.createProject(prj)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+func (mv *projectsMaterializedView) UpdateShipyard(projectName string, shipyardContent string) error {
+	existingProject, err := mv.GetProject(projectName)
+	if err != nil {
+		return err
+	}
+
+	existingProject.Shipyard = shipyardContent
+
+	return mv.updateProject(existingProject)
+}
+
+func (mv *projectsMaterializedView) GetProjects() ([]*models.ExpandedProject, error) {
+	return mv.ProjectRepo.GetProjects()
+}
+
+func (mv *projectsMaterializedView) GetProject(projectName string) (*models.ExpandedProject, error) {
+	return mv.ProjectRepo.GetProject(projectName)
+}
+
+func (mv *projectsMaterializedView) DeleteProject(projectName string) error {
+	return mv.ProjectRepo.DeleteProject(projectName)
+}
+
+func (mv *projectsMaterializedView) CreateStage(project string, stage string) error {
+	fmt.Println("Adding stage " + stage + " to project " + project)
+	prj, err := mv.GetProject(project)
+
+	if err != nil {
+		fmt.Sprintf("Could not add stage %s to project %s : %s\n", stage, project, err.Error())
+		return err
+	}
+
+	stageAlreadyExists := false
+	for _, stg := range prj.Stages {
+		if stg.StageName == stage {
+			stageAlreadyExists = true
+			break
+		}
+	}
+
+	if stageAlreadyExists {
+		fmt.Println("Stage " + stage + " already exists in project " + project)
+		return nil
+	}
+
+	prj.Stages = append(prj.Stages, &models.ExpandedStage{
+		Services:  []*models.ExpandedService{},
+		StageName: stage,
+	})
+
+	err = mv.updateProject(prj)
+	if err != nil {
+		return err
+	}
+
+	fmt.Println("Added stage " + stage + " to project " + project)
+	return nil
+}
+
+func (mv *projectsMaterializedView) createProject(prj *models.Project) error {
+	expandedProject := &models.ExpandedProject{
+		CreationDate: time.Now().String(),
+		GitRemoteURI: prj.GitRemoteURI,
+		GitUser:      prj.GitUser,
+		ProjectName:  prj.ProjectName,
+		Shipyard:     "",
+		Stages:       nil,
+	}
+
+	err := mv.ProjectRepo.CreateProject(expandedProject)
+	if err != nil {
+		fmt.Println("Could not create project " + prj.ProjectName + ": " + err.Error())
+		return err
+	}
+	return nil
+}
+
+func (mv *projectsMaterializedView) updateProject(prj *models.ExpandedProject) error {
+	return mv.ProjectRepo.UpdateProject(prj)
+}
+
+func (mv *projectsMaterializedView) DeleteStage(project string, stage string) error {
+	fmt.Println("Deleting stage " + stage + " from project " + project)
+	prj, err := mv.GetProject(project)
+
+	if err != nil {
+		fmt.Sprintf("Could not delete stage %s from project %s : %s\n", stage, project, err.Error())
+		return err
+	}
+
+	stageIndex := -1
+
+	for idx, stg := range prj.Stages {
+		if stg.StageName == stage {
+			stageIndex = idx
+			break
+		}
+	}
+	if stageIndex < 0 {
+		return nil
+	}
+
+	copy(prj.Stages[stageIndex:], prj.Stages[stageIndex+1:])
```

Comment:
A short version for deleting an element of a slice would be
`prj.Stages = append(proj.Stages[:stageIndex], proj.Stages[stageIndex+1:]...)`